### DISCUSSION TITLE
Roadmap (short-term)

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/llama.cpp#457] [Announcements]
These will be the priorities for the next few days:

- [x] **Reduce inference memory usage via [ggml scratch buffers](https://github.com/ggerganov/whisper.cpp/pull/431), no hardcoded memory buffer sizes and support infinite interactive mode**

  I know how to fix this and this is important since the GH issues are being flooded with complaints about seg faults and crashes

- [x] **Finalize SIMD accelerated quantization and merge `ggml` back in the parent repo**:
  - [x] AVX
    - [x] `quantize_row_q4_0()`
    - [x] `quantize_row_q4_1()`
    - [x] `dequantize_row_q4_0()`
    - [x] `dequantize_row_q4_1()`
  - [x] ARM_NEON
    - [x] `quantize_row_q4_0()`
    - [x] `quantize_row_q4_1()`
    - [x] `dequantize_row_q4_0()`
    - [x] `dequantize_row_q4_1()`

  [I suspect this could improve performance for prompt batch processing](https://github.com/ggerganov/llama.cpp/discussions/229#discussioncomment-5409647)

- [x] **[Deprecate `ggml_vec_mad_xxx()` routines and simplify `ggml_forward_mul_mat_xxx()`](https://github.com/ggerganov/llama.cpp/issues/441)**

  This should lead to some significant code reduction in `ggml.c`

- [x]  **Separate the perplexity computation from `main.cpp` into standalone example program called `perplexity`**

- [x] **Move `main.cpp` into a standalone example program and move `utils.h`/`utils.cpp` into `./examples` to be shared by all examples**

- [ ] **[Add `llama_state`](https://github.com/ggerganov/llama.cpp/pull/370#discussion_r1144235066) to allow parallel text generation sessions with a single model**

  I will do this in a similar way it is done in `whisper.cpp`

- [ ] **Extend `llama_state` to support loading individual model tensors. Needed for [LoRA personalities support](https://github.com/ggerganov/llama.cpp/issues/399)**
 
- [ ] **[Add 2-bit integer quantization](https://github.com/ggerganov/llama.cpp/issues/456)**

---

When the above things are ready we will have a good foundation to start porting more models and create more example applications to demonstrate the usage of `ggml`.

New roadmap: https://github.com/ggerganov/llama.cpp/discussions/784

**Comment by niansa** [comment]
Are these sorted top-priority first?

**Reply by Green-Sky** [reply]
more or less i think

**Comment by jasontitus** [comment]
Can we update the README with usage of the new perplexity tool since the 'main --perplexity' way stopped working?

**Reply by Green-Sky** [reply]
it's now it's own program
(note: make it link to examples/perplexity/README.md and fill in the doc there)

**Comment by wildcat7534** [comment]
What's append, is the projet become re activate  later ? u_u
